Error opening SFTP connection with public key

## Issue explanation (write below this line)

I get an error when trying to open a SFTP connection with a public key (which was working fine with a previous version of Amaze).

## Exception
* __App Name:__ Administrador de Archivos Amaze
* __Package:__ com.amaze.filemanager
* __Version:__ 3.8.2
* __User Action:__ UI Error
* __Request:__ Application crash
* __OS:__ Linux Android 13 - 33
* __Device:__ bluejay
* __Model:__ Pixel 6a
* __Product:__ bluejay
<details><summary><b>Crash log </b></summary><p>

```
io.reactivex.exceptions.CompositeException: 2 exceptions occurred. 
	at io.reactivex.internal.operators.maybe.MaybeCallbackObserver.onError(MaybeCallbackObserver.java:85)
	at io.reactivex.internal.operators.maybe.MaybeSubscribeOn$SubscribeOnMaybeObserver.onError(MaybeSubscribeOn.java:96)
	at io.reactivex.internal.operators.maybe.MaybeFromCallable.subscribeActual(MaybeFromCallable.java:50)
	at io.reactivex.Maybe.subscribe(Maybe.java:4290)
	at io.reactivex.internal.operators.maybe.MaybeSubscribeOn$SubscribeTask.run(MaybeSubscribeOn.java:54)
	at io.reactivex.Scheduler$DisposeTask.run(Scheduler.java:578)
	at io.reactivex.internal.schedulers.ScheduledRunnable.run(ScheduledRunnable.java:66)
	at io.reactivex.internal.schedulers.ScheduledRunnable.call(ScheduledRunnable.java:57)
	at java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:307)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1137)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:637)
	at java.lang.Thread.run(Thread.java:1012)
  ComposedException 1 :
	net.schmizz.sshj.transport.TransportException: [HOST_KEY_NOT_VERIFIABLE] Could not verify `ssh-ed25519` host key with fingerprint `28:f4:14:1f:6f:80:7c:89:7a:d9:01:72:17:0c:90:0a` for `172.27.7.33` on port 22
		at net.schmizz.sshj.transport.KeyExchanger.verifyHost(KeyExchanger.java:220)
		at net.schmizz.sshj.transport.KeyExchanger.handle(KeyExchanger.java:390)
		at net.schmizz.sshj.transport.TransportImpl.handle(TransportImpl.java:474)
		at net.schmizz.sshj.transport.Decoder.decode(Decoder.java:113)
		at net.schmizz.sshj.transport.Decoder.received(Decoder.java:200)
		at net.schmizz.sshj.transport.Reader.run(Reader.java:60)
	Caused by: java.lang.RuntimeException: Can't create handler inside thread Thread[RxCachedThreadScheduler-1,5,main] that has not called Looper.prepare()
		at android.os.Handler.<init>(Handler.java:227)
		at android.os.Handler.<init>(Handler.java:129)
		at android.app.Dialog.<init>(Dialog.java:138)
		at android.app.AlertDialog.<init>(AlertDialog.java:204)
		at android.app.AlertDialog$Builder.create(AlertDialog.java:1105)
		at android.app.AlertDialog$Builder.show(AlertDialog.java:1130)
		at com.amaze.filemanager.asynchronous.asynctasks.ssh.SshAuthenticationTask.onError(SshAuthenticationTask.kt:78)
		at com.amaze.filemanager.filesystem.ftp.NetCopyClientConnectionPool.createSshClientInternal$lambda-7(NetCopyClientConnectionPool.kt:319)
		at com.amaze.filemanager.filesystem.ftp.NetCopyClientConnectionPool.$r8$lambda$ehN52awM7RuNe3l_m2OrcGDs2Ac(Unknown Source:0)
		at com.amaze.filemanager.filesystem.ftp.NetCopyClientConnectionPool$$ExternalSyntheticLambda2.accept(Unknown Source:6)
		at io.reactivex.internal.operators.maybe.MaybeCallbackObserver.onError(MaybeCallbackObserver.java:82)
		at io.reactivex.internal.operators.maybe.MaybeSubscribeOn$SubscribeOnMaybeObserver.onError(MaybeSubscribeOn.java:96)
		at io.reactivex.internal.operators.maybe.MaybeFromCallable.subscribeActual(MaybeFromCallable.java:50)
		at io.reactivex.Maybe.subscribe(Maybe.java:4290)
		at io.reactivex.internal.operators.maybe.MaybeSubscribeOn$SubscribeTask.run(MaybeSubscribeOn.java:54)
		at io.reactivex.Scheduler$DisposeTask.run(Scheduler.java:578)
		at io.reactivex.internal.schedulers.ScheduledRunnable.run(ScheduledRunnable.java:66)
		at io.reactivex.internal.schedulers.ScheduledRunnable.call(ScheduledRunnable.java:57)
		at java.util.concurrent.FutureTask.run(FutureTask.java:264)
		at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:307)
		at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1137)
		at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:637)
		at java.lang.Thread.run(Thread.java:1012)
  ComposedException 2 :
	java.lang.RuntimeException: Can't create handler inside thread Thread[RxCachedThreadScheduler-1,5,main] that has not called Looper.prepare()
		at android.os.Handler.<init>(Handler.java:227)
		at android.os.Handler.<init>(Handler.java:129)
		at android.app.Dialog.<init>(Dialog.java:138)
		at android.app.AlertDialog.<init>(AlertDialog.java:204)
		at android.app.AlertDialog$Builder.create(AlertDialog.java:1105)
		at android.app.AlertDialog$Builder.show(AlertDialog.java:1130)
		at com.amaze.filemanager.asynchronous.asynctasks.ssh.SshAuthenticationTask.onError(SshAuthenticationTask.kt:78)
		at com.amaze.filemanager.filesystem.ftp.NetCopyClientConnectionPool.createSshClientInternal$lambda-7(NetCopyClientConnectionPool.kt:319)
		at com.amaze.filemanager.filesystem.ftp.NetCopyClientConnectionPool.$r8$lambda$ehN52awM7RuNe3l_m2OrcGDs2Ac(Unknown Source:0)
		at com.amaze.filemanager.filesystem.ftp.NetCopyClientConnectionPool$$ExternalSyntheticLambda2.accept(Unknown Source:6)
		at io.reactivex.internal.operators.maybe.MaybeCallbackObserver.onError(MaybeCallbackObserver.java:82)
		at io.reactivex.internal.operators.maybe.MaybeSubscribeOn$SubscribeOnMaybeObserver.onError(MaybeSubscribeOn.java:96)
		at io.reactivex.internal.operators.maybe.MaybeFromCallable.subscribeActual(MaybeFromCallable.java:50)
		at io.reactivex.Maybe.subscribe(Maybe.java:4290)
		at io.reactivex.internal.operators.maybe.MaybeSubscribeOn$SubscribeTask.run(MaybeSubscribeOn.java:54)
		at io.reactivex.Scheduler$DisposeTask.run(Scheduler.java:578)
		at io.reactivex.internal.schedulers.ScheduledRunnable.run(ScheduledRunnable.java:66)
		at io.reactivex.internal.schedulers.ScheduledRunnable.call(ScheduledRunnable.java:57)
		at java.util.concurrent.FutureTask.run(FutureTask.java:264)
		at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:307)
		at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1137)
		at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:637)
		at java.lang.Thread.run(Thread.java:1012)


```
</details>
<hr>

Same here. I'm using RSA key (4096 bis) and passphrase. My logs:
## Exception
* __App Name:__ Amaze File Manager
* __Package:__ com.amaze.filemanager
* __Version:__ 3.8.5
* __User Action:__ UI Error
* __Request:__ Application crash
* __OS:__ Linux Android 13 - 33
* __Device:__ OnePlus6
* __Model:__ ONEPLUS A6003
* __Product:__ OnePlus6
<details><summary><b>Crash log </b></summary><p>

```
java.lang.RuntimeException: An error occurred while executing doInBackground()
	at android.os.AsyncTask$4.done(AsyncTask.java:415)
	at java.util.concurrent.FutureTask.finishCompletion(FutureTask.java:381)
	at java.util.concurrent.FutureTask.setException(FutureTask.java:250)
	at java.util.concurrent.FutureTask.run(FutureTask.java:269)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1137)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:637)
	at java.lang.Thread.run(Thread.java:1012)
Caused by: java.lang.NumberFormatException: For input string: "\tmp"
	at java.lang.Integer.parseInt(Integer.java:747)
	at java.lang.Integer.parseInt(Integer.java:865)
	at com.amaze.filemanager.filesystem.ftp.NetCopyClientConnectionPool$ConnectionInfo.<init>(NetCopyClientConnectionPool.kt:416)
	at com.amaze.filemanager.filesystem.ftp.NetCopyClientConnectionPool.createSshClient(NetCopyClientConnectionPool.kt:252)
	at com.amaze.filemanager.filesystem.ftp.NetCopyClientConnectionPool.access$createSshClient(NetCopyClientConnectionPool.kt:48)
	at com.amaze.filemanager.filesystem.ftp.NetCopyClientConnectionPool$createNetCopyClient$1.invoke(NetCopyClientConnectionPool.kt:175)
	at com.amaze.filemanager.filesystem.ftp.NetCopyClientConnectionPool$createNetCopyClient$1.invoke(NetCopyClientConnectionPool.kt:173)
	at com.amaze.filemanager.filesystem.ftp.NetCopyClientConnectionPool.getConnection(NetCopyClientConnectionPool.kt:88)
	at com.amaze.filemanager.filesystem.ftp.NetCopyClientUtils.execute(NetCopyClientUtils.kt:83)
	at com.amaze.filemanager.filesystem.HybridFile.forEachChildrenFile(HybridFile.java:938)
	at com.amaze.filemanager.asynchronous.asynctasks.LoadFilesListTask.listSftp(LoadFilesListTask.java:626)
	at com.amaze.filemanager.asynchronous.asynctasks.LoadFilesListTask.doInBackground(LoadFilesListTask.java:157)
	at com.amaze.filemanager.asynchronous.asynctasks.LoadFilesListTask.doInBackground(LoadFilesListTask.java:85)
	at android.os.AsyncTask$3.call(AsyncTask.java:394)
	at java.util.concurrent.FutureTask.run(FutureTask.java:264)
	... 3 more

```
</details>
<hr>

--- a/app/src/main/java/com/amaze/filemanager/asynchronous/services/CopyService.java
+++ b/app/src/main/java/com/amaze/filemanager/asynchronous/services/CopyService.java
@@ -479,7 +479,9 @@ public class CopyService extends AbstractProgressiveService {
         if (progressHandler.getCancelled()) return;
         if (sourceFile.isDirectory()) {
 
-          if (!targetFile.exists()) targetFile.mkdir(c);
+          if (!targetFile.exists()) {
+            targetFile.mkdir(c);
+          }
 
           // various checks
           // 1. source file and target file doesn't end up in loop

--- a/app/src/main/java/com/amaze/filemanager/filesystem/HybridFile.java
+++ b/app/src/main/java/com/amaze/filemanager/filesystem/HybridFile.java
@@ -76,7 +76,6 @@ import com.amaze.filemanager.filesystem.root.DeleteFileCommand;
 import com.amaze.filemanager.filesystem.root.ListFilesCommand;
 import com.amaze.filemanager.filesystem.ssh.SFtpClientTemplate;
 import com.amaze.filemanager.filesystem.ssh.SshClientSessionTemplate;
-import com.amaze.filemanager.filesystem.ssh.SshClientTemplate;
 import com.amaze.filemanager.filesystem.ssh.SshClientUtils;
 import com.amaze.filemanager.filesystem.ssh.Statvfs;
 import com.amaze.filemanager.ui.activities.MainActivity;
@@ -1198,14 +1197,13 @@ public class HybridFile {
     switch (mode) {
       case SFTP:
         return NetCopyClientUtils.INSTANCE.execute(
-            new SshClientTemplate<OutputStream>(path, false) {
+            new SFtpClientTemplate<OutputStream>(getPath(), false) {
+              @Nullable
               @Override
-              public OutputStream executeWithSSHClient(@NonNull final SSHClient ssh)
-                  throws IOException {
-                final SFTPClient client = ssh.newSFTPClient();
+              public OutputStream execute(@NonNull SFTPClient client) throws IOException {
                 final RemoteFile rf =
                     client.open(
-                        NetCopyClientUtils.INSTANCE.extractRemotePathFrom(path),
+                        NetCopyClientUtils.INSTANCE.extractRemotePathFrom(getPath()),
                         EnumSet.of(
                             net.schmizz.sshj.sftp.OpenMode.WRITE,
                             net.schmizz.sshj.sftp.OpenMode.CREAT));
@@ -1287,7 +1285,7 @@ public class HybridFile {
     if (isSftp()) {
       final Boolean executionReturn =
           NetCopyClientUtils.INSTANCE.<SSHClient, Boolean>execute(
-              new SFtpClientTemplate<Boolean>(path, false) {
+              new SFtpClientTemplate<Boolean>(path, true) {
                 @Override
                 public Boolean execute(SFTPClient client) throws IOException {
                   try {
@@ -1412,17 +1410,16 @@ public class HybridFile {
 
   public void mkdir(Context context) {
     if (isSftp()) {
-      NetCopyClientUtils.INSTANCE.execute(
-          new SFtpClientTemplate<Void>(path, false) {
+      NetCopyClientUtils.INSTANCE.<SSHClient, Boolean>execute(
+          new SFtpClientTemplate<Boolean>(path, true) {
             @Override
-            public Void execute(@NonNull SFTPClient client) {
+            public Boolean execute(@NonNull SFTPClient client) {
               try {
                 client.mkdir(NetCopyClientUtils.INSTANCE.extractRemotePathFrom(path));
               } catch (IOException e) {
                 LOG.error("Error making directory over SFTP", e);
               }
-              // FIXME: anything better than throwing a null to make Rx happy?
-              return null;
+              return true;
             }
           });
     } else if (isFtp()) {

--- a/app/src/main/java/com/amaze/filemanager/ui/dialogs/SftpConnectDialog.kt
+++ b/app/src/main/java/com/amaze/filemanager/ui/dialogs/SftpConnectDialog.kt
@@ -29,6 +29,7 @@ import android.content.Intent
 import android.net.Uri
 import android.os.Bundle
 import android.text.Editable
+import android.text.InputFilter
 import android.text.TextUtils
 import android.text.TextWatcher
 import android.view.LayoutInflater
@@ -37,6 +38,7 @@ import android.widget.AdapterView
 import android.widget.ArrayAdapter
 import androidx.activity.result.contract.ActivityResultContracts
 import androidx.core.text.isDigitsOnly
+import androidx.core.widget.addTextChangedListener
 import androidx.fragment.app.DialogFragment
 import com.afollestad.materialdialogs.DialogAction
 import com.afollestad.materialdialogs.MaterialDialog
@@ -104,6 +106,20 @@ class SftpConnectDialog : DialogFragment() {
         const val ARG_KEYPAIR_NAME = "keypairName"
 
         private val VALID_PORT_RANGE = IntRange(1, 65535)
+
+        // Loosely referenced from https://dwheeler.com/essays/fixing-unix-linux-filenames.html
+        private const val pathBlockedChars = "*?<>|\\"
+
+        private val defaultPathCharFilter =
+            InputFilter { source, _, _, _, _, _ ->
+                if (source.isNotEmpty() && source.isNotBlank() &&
+                    pathBlockedChars.contains(source)
+                ) {
+                    ""
+                } else {
+                    null
+                }
+            }
     }
 
     lateinit var ctx: WeakReference<Context>
@@ -190,7 +206,7 @@ class SftpConnectDialog : DialogFragment() {
                 passwordET.setText("")
             }
         }
-
+        defaultPathET.filters = arrayOf(defaultPathCharFilter)
         // If it's new connection setup, set some default values
         // Otherwise, use given Bundle instance for filling in the blanks
         if (!edit) {

